File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

"""
         1
       2     
    4     5
  12   6     7
     8  9  10 
          11
"""

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.diameter = 0
        self.dfs(root)
        return self.diameter
    
    def dfs(self, node):
        # base case:
        if node == None:
            return 0
		# recursive cases
        left_height = self.dfs(node.left)
        right_height = self.dfs(node.right)
        self.diameter = max(self.diameter,left_height + right_height )
        return max(left_height,right_height) + 1

        return min(self.dfsMin(node.left), self.dfsMin(node.right)) + 1
    
    

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int longest = 0;
    
    public int diameterOfBinaryTree(TreeNode root) {
        recur(root);
        
        return longest;
    }
    
    public int recur(TreeNode root){
        if(root == null)
            return 0;
        
        int leftMax = recur(root.left);
        int rightMax = recur(root.right);
        
        longest = Math.max(longest, leftMax + rightMax);
        
        return Math.max(leftMax, rightMax) + 1;
    }
}